File: android/src/main/java/me/carda/awesome_notifications/awesome_notifications_android_core/broadcasters/receivers/AwesomeEventsReceiver.java
package me.carda.awesome_notifications.awesome_notifications_android_core.broadcasters.receivers;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import me.carda.awesome_notifications.awesome_notifications_android_core.AwesomeNotifications;
import me.carda.awesome_notifications.awesome_notifications_android_core.Definitions;
import me.carda.awesome_notifications.awesome_notifications_android_core.listeners.AwesomeActionEventListener;
import me.carda.awesome_notifications.awesome_notifications_android_core.listeners.AwesomeNotificationEventListener;
import me.carda.awesome_notifications.awesome_notifications_android_core.exceptions.AwesomeNotificationException;
import me.carda.awesome_notifications.awesome_notifications_android_core.models.returnedData.ActionReceived;
import me.carda.awesome_notifications.awesome_notifications_android_core.models.returnedData.NotificationReceived;
import me.carda.awesome_notifications.awesome_notifications_android_core.utils.StringUtils;

public class AwesomeEventsReceiver {

    public static String TAG = "AwesomeIntentFiltersReceiver";

    // ************** SINGLETON PATTERN ***********************

    @SuppressLint("StaticFieldLeak")
    private static AwesomeEventsReceiver instance;

    private AwesomeEventsReceiver(){
    }

    public static AwesomeEventsReceiver getInstance() {
        if (instance == null)
            instance = new AwesomeEventsReceiver();
        return instance;
    }

    // ********************************************************

    /// **************  OBSERVER PATTERN  *********************

    static List<AwesomeNotificationEventListener> notificationEventListeners = new ArrayList<>();
    public AwesomeEventsReceiver subscribeOnNotificationEvents(AwesomeNotificationEventListener listener) {
        notificationEventListeners.add(listener);
        return this;
    }
    public AwesomeEventsReceiver unsubscribeOnNotificationEvents(AwesomeNotificationEventListener listener) {
        notificationEventListeners.remove(listener);
        return this;
    }
    private void notifyNotificationEvent(String eventName, NotificationReceived notificationReceived) {
        for (AwesomeNotificationEventListener listener : notificationEventListeners)
            listener.onNewNotificationReceived(eventName, notificationReceived);
    }

    // ********************************************************

    static List<AwesomeActionEventListener> notificationActionListeners = new ArrayList<>();
    public AwesomeEventsReceiver subscribeOnActionEvents(AwesomeActionEventListener listener) {
        notificationActionListeners.add(listener);
        return this;
    }
    public AwesomeEventsReceiver unsubscribeOnActionEvents(AwesomeActionEventListener listener) {
        notificationActionListeners.remove(listener);
        return this;
    }
    private void notifyActionEvent(String eventName, ActionReceived actionReceived) {
        for (AwesomeActionEventListener listener : notificationActionListeners)
            listener.onNewActionReceived(eventName, actionReceived);
    }

    // ********************************************************

    public void addAwesomeNotificationEvent(Context context, String eventName, NotificationReceived notificationReceived){

        if(notificationEventListeners.isEmpty())
            return;

        try {
            switch (eventName) {

                case Definitions.BROADCAST_CREATED_NOTIFICATION:
                    onBroadcastNotificationCreated(context, notificationReceived);
                    return;

                case Definitions.BROADCAST_DISPLAYED_NOTIFICATION:
                    onBroadcastNotificationDisplayed(context, notificationReceived);
                    return;

                default:
                    if (AwesomeNotifications.debug)
                        Log.d(TAG, "Received unknown notification event: " + (
                                StringUtils.isNullOrEmpty(eventName) ? "empty" : eventName));

            }
        } catch (Exception e) {
            if (AwesomeNotifications.debug)
                Log.d(TAG, String.format("%s", e.getMessage()));
            e.printStackTrace();
        }
    }

    public void addAwesomeActionEvent(Context context, String eventName, ActionReceived actionReceived){

        try {
            switch (eventName) {

                case Definitions.BROADCAST_DEFAULT_ACTION:
                    onBroadcastDefaultActionNotification(context, actionReceived);
                    return;

                case Definitions.BROADCAST_DISMISSED_NOTIFICATION:
                    onBroadcastNotificationDismissed(context, actionReceived);
                    return;

                case Definitions.BROADCAST_SILENT_ACTION:
                    onBroadcastSilentActionNotification(context, actionReceived);
                    return;

                case Definitions.BROADCAST_BACKGROUND_ACTION:
                    onBroadcastBackgroundActionNotification(context, actionReceived);
                    return;

                default:
                    if (AwesomeNotifications.debug)
                        Log.d(TAG, "Received unknown action event: " + (
                                StringUtils.isNullOrEmpty(eventName) ? "empty" : eventName));

            }
        } catch (Exception e) {
            if (AwesomeNotifications.debug)
                Log.d(TAG, String.format("%s", e.getMessage()));
            e.printStackTrace();
        }
    }

    /// ***********************************

    private void onBroadcastNotificationCreated(Context context, NotificationReceived notificationReceived) throws AwesomeNotificationException {
        try {
            notificationReceived.validate(context);

            if (AwesomeNotifications.debug)
                Log.d(TAG, "Notification created");

            notifyNotificationEvent(Definitions.CHANNEL_METHOD_NOTIFICATION_CREATED, notificationReceived);

        } catch (Exception e) {
            throw new AwesomeNotificationException(e.getMessage());
        }
    }

    private void onBroadcastNotificationDisplayed(Context context, NotificationReceived notificationReceived) throws AwesomeNotificationException {
        try {
            notificationReceived.validate(context);

            if (AwesomeNotifications.debug)
                Log.d(TAG, "Notification displayed");

            notifyNotificationEvent(Definitions.CHANNEL_METHOD_NOTIFICATION_DISPLAYED, notificationReceived);

        } catch (Exception e) {
            throw new AwesomeNotificationException(e.getMessage());
        }
    }

    private void onBroadcastDefaultActionNotification(Context context, ActionReceived actionReceived) {
        try {
            actionReceived.validate(context);

            if(AwesomeNotifications.debug)
                Log.d(TAG, "Notification action received");

            notifyActionEvent(Definitions.CHANNEL_METHOD_DEFAULT_ACTION, actionReceived);

        } catch (Exception e) {
            if(AwesomeNotifications.debug)
                Log.d(TAG, String.format("%s", e.getMessage()));
            e.printStackTrace();
        }
    }

    private void onBroadcastNotificationDismissed(Context context, ActionReceived actionReceived) throws AwesomeNotificationException {
        try {
            actionReceived.validate(context);

            if (AwesomeNotifications.debug)
                Log.d(TAG, "Notification dismissed");

            notifyActionEvent(Definitions.CHANNEL_METHOD_NOTIFICATION_DISMISSED, actionReceived);

        } catch (Exception e) {
            throw new AwesomeNotificationException(e.getMessage());
        }
    }

    private void onBroadcastSilentActionNotification(Context context, ActionReceived actionReceived) {
        try {
            actionReceived.validate(context);

            if(AwesomeNotifications.debug)
                Log.d(TAG, "Notification silent action received");

            notifyActionEvent(Definitions.CHANNEL_METHOD_SILENT_ACTION, actionReceived);

        } catch (Exception e) {
            if(AwesomeNotifications.debug)
                Log.d(TAG, String.format("%s", e.getMessage()));
            e.printStackTrace();
        }
    }

    private void onBroadcastBackgroundActionNotification(Context context, ActionReceived actionReceived) throws AwesomeNotificationException {
        try {
            actionReceived.validate(context);

            if (AwesomeNotifications.debug)
                Log.d(TAG, "Notification action received");

            notifyActionEvent(Definitions.CHANNEL_METHOD_SILENT_ACTION, actionReceived);

        } catch (Exception e) {
            throw new AwesomeNotificationException(e.getMessage());
        }
    }
}

File: ios/Classes/lib/awesome_notifications_core/broadcasters/receivers/AwesomeEventsReceiver.swift
//
//  AwesomeEventsReceiver.swift
//  awesome_notifications
//
//  Created by CardaDev on 02/02/22.
//

import Foundation

class AwesomeEventsReceiver {
    
    private let TAG = "AwesomeEventsReceiver"
    
    
    // **************************** SINGLETON PATTERN *************************************
    
    static var instance:AwesomeEventsReceiver?
    public static var shared:AwesomeEventsReceiver {
        get {
            AwesomeEventsReceiver.instance =
                AwesomeEventsReceiver.instance ?? AwesomeEventsReceiver()
            return AwesomeEventsReceiver.instance!
        }
    }
    private init(){}
    
    
    // **************************** OBSERVER PATTERN **************************************
    
    private lazy var notificationEventListeners = [AwesomeNotificationEventListener]()
    
    public func subscribeOnNotificationEvents(listener:AwesomeNotificationEventListener) -> Self {
        notificationEventListeners.append(listener)
        return self
    }
    
    public func unsubscribeOnNotificationEvents(listener:AwesomeNotificationEventListener) -> Self {
        if let index = notificationEventListeners.firstIndex(where: {$0 === listener}) {
            notificationEventListeners.remove(at: index)
        }
        return self
    }
    
    private func notifyNotificationEvent(
        named eventName: String,
        with notificationReceived: NotificationReceived
    ){
        for listener in notificationEventListeners {
            listener.onNewNotificationReceived(
                eventName: eventName,
                notificationReceived: notificationReceived)
        }
    }
    
    
    // **************************** OBSERVER PATTERN **************************************
    
    private lazy var actionEventListeners = [AwesomeActionEventListener]()
    
    public func subscribeOnActionEvents(listener:AwesomeActionEventListener) {
        actionEventListeners.append(listener)
    }
    
    public func unsubscribeOnActionEvents(listener:AwesomeActionEventListener) {
        if let index = actionEventListeners.firstIndex(where: {$0 === listener}) {
            actionEventListeners.remove(at: index)
        }
    }
    
    private func notifyActionEvent(
        named eventName: String,
        with actionReceived: ActionReceived
    ){
        for listener in actionEventListeners {
            listener.onNewActionReceived(
                fromEventNamed: eventName,
                withActionReceived: actionReceived)
        }
    }
    
    // **************************** OBSERVER PATTERN **************************************
    
    
    public func addNotificationEvent(
        named eventName: String,
        with notificationReceived: NotificationReceived
    ){
        
        if notificationEventListeners.isEmpty {
            return
        }
        
        do {
            switch eventName {
                
                case Definitions.BROADCAST_CREATED_NOTIFICATION:
                    try onBroadcast(notificationCreated: notificationReceived)
                    return
                    
                case Definitions.BROADCAST_DISPLAYED_NOTIFICATION:
                    try onBroadcast(notificationDisplayed: notificationReceived)
                    return
                    
                default:
                    if AwesomeNotifications.debug {
                        Log.d(TAG, "Received unknown notification event: '\(eventName)'")
                    }
            }
        } catch {
            Log.d(TAG, "\(error).")
        }
    }
    
    public func addActionEvent(
        named eventName: String,
        with actionReceived: ActionReceived
    ){
        do {
            switch eventName {
                
                case Definitions.BROADCAST_DEFAULT_ACTION:
                    try onBroadcast(defaultAction: actionReceived)
                    return
                    
                case Definitions.BROADCAST_DISMISSED_NOTIFICATION:
                    try onBroadcast(dismissAction: actionReceived)
                    return
                    
                case Definitions.BROADCAST_SILENT_ACTION:
                    try onBroadcast(silentAction: actionReceived)
                    return
                    
                case Definitions.BROADCAST_BACKGROUND_ACTION:
                    try onBroadcast(backgroundAction: actionReceived)
                    return
                    
                default:
                    if AwesomeNotifications.debug {
                        Log.d(TAG, "Received unknown notification event: '\(eventName)'")
                    }
            }
        }
        catch {
            Log.d(TAG, "\(error).")
        }
    }
    
    private func onBroadcast(notificationCreated notificationReceived: NotificationReceived) throws {
        try notificationReceived.validate()
        
        if AwesomeNotifications.debug {
            Log.d(TAG, "Notification created")
        }
        
        notifyNotificationEvent(
            named: Definitions.EVENT_NOTIFICATION_CREATED,
            with: notificationReceived)
    }
    
    private func onBroadcast(notificationDisplayed notificationReceived: NotificationReceived) throws {
        try notificationReceived.validate()
        
        if AwesomeNotifications.debug {
            Log.d(TAG, "Notification displayed")
        }
        
        notifyNotificationEvent(
            named: Definitions.EVENT_NOTIFICATION_DISPLAYED,
            with: notificationReceived)
    }
    
    private func onBroadcast(defaultAction actionReceived: ActionReceived) throws {
        try actionReceived.validate()
        
        if AwesomeNotifications.debug {
            Log.d(TAG, "Notification action received")
        }
        
        notifyActionEvent(
            named: Definitions.EVENT_DEFAULT_ACTION,
            with: actionReceived)
    }
    
    private func onBroadcast(dismissAction actionReceived: ActionReceived) throws {
        try actionReceived.validate()
        
        if AwesomeNotifications.debug {
            Log.d(TAG, "Notification dismissed")
        }
        
        notifyActionEvent(
            named: Definitions.EVENT_NOTIFICATION_DISMISSED,
            with: actionReceived)
    }
    
    private func onBroadcast(silentAction actionReceived: ActionReceived) throws {
        try actionReceived.validate()
        
        if AwesomeNotifications.debug {
            Log.d(TAG, "Silent action received")
        }
        
        notifyActionEvent(
            named: Definitions.EVENT_SILENT_ACTION,
            with: actionReceived)
    }
    
    private func onBroadcast(backgroundAction actionReceived: ActionReceived) throws {
        try actionReceived.validate()
        
        if AwesomeNotifications.debug {
            Log.d(TAG, "Background silent action received")
        }
        
        notifyActionEvent(
            named: Definitions.EVENT_SILENT_ACTION,
            with: actionReceived)
    }
}
